internal/redact/redact.go

```diff
@@ -0,0 +1,218 @@
+//
+//	type userErr struct{ name string }
+//
+//	func (e *userErr) Error() string {
+//		return fmt.Sprintf("user %s not found", e.name)
+//	}
+//
+//	func (e *userErr) Redact() string {
+//		return fmt.Sprintf("user %x not found", sha256.Sum256([]byte(e.name)))
+//	}
+//
+//	func main() {
+//		err := &userErr{name: "Alex"}
+//		fmt.Println(err)
+//		// user Alex not found
+
+//		fmt.Println(Error(err))
+//		// user db74c940d447e877d119df613edd2700c4a84cd1cf08beb7cbc319bcfaeab97a not found
+//	}
+//
+// The [Errorf] function creates redactable errors that retain their literal
+// format text, but redact any arguments. The format string spec is identical
+// to that of [fmt.Errorf]. Calling [Safe] on an [Errorf] argument will include
+// it in the redacted message.
+//
+//	name := "Alex"
+//	id := 5
+//	err := Errorf("error getting user %s with ID %d", name, Safe(id))
+//
+//	fmt.Println(err)
+//	// error getting user Alex with ID 5
+//
+//	fmt.Println(Error(err))
+//	// error getting user <redacted string> with ID 5
+package redact
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+)
+
+// Error returns a redacted error that wraps err. If err has a Redact() string
+// method, then Error uses it for the redacted error message. Otherwise, Error
+// recursively redacts each wrapped error, joining them with ": " to create the
+// final error message. If it encounters an error that has a Redact() method,
+// then it appends the result of Redact() to the message and stops unwrapping.
+func Error(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	switch t := err.(type) {
+	case *safeError:
+		return t.redacted
+	case redactor:
+		return &redactedError{
+			msg:     t.Redact(),
+			wrapped: err,
+		}
+	default:
+		msg := placeholder(err)
+		wrapped := err
+		for {
+			wrapped = errors.Unwrap(wrapped)
+			if wrapped == nil {
+				break
+			}
+			if redactor, ok := wrapped.(redactor); ok {
+				msg += ": " + redactor.Redact()
+				break
+			}
+			msg += ": " + placeholder(wrapped)
+		}
+		return &redactedError{
+			msg:     msg,
+			wrapped: err,
+		}
+	}
+}
+
+// Errorf creates a redactable error that has an error string identical to that
+// of a [fmt.Errorf] error. Calling [Redact] on the result will redact all
+// format arguments from the error message instead of redacting the entire
+// string.
+//
+// When redacting the error string, Errorf replaces arguments that implement a
+// Redact() string method with the result of that method. To include an
+// argument as-is in the redacted error, first call [Safe]. For example:
+//
+//	username := "bob"
+//	Errorf("cannot find user %s", username).Error()
+//	// cannot find user <redacted string>
+//
+//	Errorf("cannot find user %s", Safe(username)).Error()
+//	// cannot find user bob
+func Errorf(format string, a ...any) error {
+	// Capture a stack trace.
+	safeErr := &safeError{
+		callers: make([]uintptr, 32),
+	}
+	n := runtime.Callers(2, safeErr.callers)
+	safeErr.callers = safeErr.callers[:n]
```

Comment:
Is this going to print stack traces correctly with out existing error handling? (using `pkg/errors` ? )

Reply:
I updated `safeError` to format its stack trace the same way `pkg/errors` does. I double checked and the output is identical.